Some caches not invalidated after parameter re-assignment

**Description**

When a parameter is updated after `Interchange` creation, some caches used by exports are not invalidated/updated. These caches were introduced in https://github.com/openforcefield/openff-interchange/pull/1122

I don't think this is a necessarily common operation but it is possible with public API points and, for better or worse, included in the [user guide](https://docs.openforcefield.org/projects/interchange/en/stable/using/collections.html#)

#1234 and #1433 are related but I don't think necessarily the same issue. I will need to double-check

Credit to @j-wags whose question [got the ball rolling](https://github.com/openforcefield/openff-interchange/pull/1434#discussion_r2856196151) on this

**Reproduction**

Here are two unit tests - the first passes and the second fails (for all inputs)

```diff
diff --git a/openff/interchange/_tests/test_issues.py b/openff/interchange/_tests/test_issues.py
index 5e3fafee..fddce7f6 100644
--- a/openff/interchange/_tests/test_issues.py
+++ b/openff/interchange/_tests/test_issues.py
@@ -262,6 +262,41 @@ def test_issue_1433_openmm(sage, valence_handler):
     assert abs(original_energy - new_energy) > 0.1
 
 
+def test_reassign_vdw(sage, water_dimer):
+    """Sanity check that re-assigning vdW parameters actually changes energies."""
+    original_interchange = sage.create_interchange(water_dimer)
+
+    original_energy = get_openmm_energies(original_interchange).energies["Nonbonded"]
+
+    for potential in original_interchange["vdW"].potentials.values():
+        potential.parameters["epsilon"] = potential.parameters["epsilon"] * 100
+
+    new_energy = get_openmm_energies(original_interchange).energies["Nonbonded"]
+
+    assert original_energy != new_energy, "Energies should be different after modifying vdW parameters"
+
+
+@pytest.mark.parametrize("valence_handler", ["Bonds", "Angles", "ProperTorsions"])
+def test_reassign_valence(sage, valence_handler):
+    """Sanity check that re-assigning valence parameters actually changes energies."""
+    topology = MoleculeWithConformer.from_smiles(
+        "c1c(cccc1)C(c2ccccc2)c3ccccc3",
+        allow_undefined_stereo=True,
+    ).to_topology()
+    original_interchange = sage.create_interchange(topology)
+
+    original_energy = get_openmm_energies(original_interchange)
+
+    print(original_interchange[valence_handler].get_system_parameters())
+    for potential in original_interchange[valence_handler].potentials.values():
+        potential.parameters["k"] = potential.parameters["k"] * 100
+
+    print(original_interchange[valence_handler].get_system_parameters())
+    new_energy = get_openmm_energies(original_interchange)
+
+    assert original_energy != new_energy, "Energies should be different after modifying valence parameters"
+
+
 @pytest.mark.parametrize("valence_handler", ["Bonds", "Angles", "ProperTorsions", "ImproperTorsions"])
 def test_issue_1234_openmm(sage, valence_handler):
     """Test that modifications to a `ForceField` object are reflected in re-creating new `Interchange`s."""

```
**Output**

```
$ pixi run -e py312openeye pytest openff/interchange/_tests/test_issues.py::test_reassign_vdw openff/interchange/_tests/test_issues.py::test_reassign_valence -q
F.FF                                                                                                                                               [100%]
======================================================================== FAILURES ========================================================================
_____________________________________________________________ test_reassign_valence[Angles] ______________________________________________________________

sage = <openff.toolkit.typing.engines.smirnoff.forcefield.ForceField object at 0x179fc8ec0>, valence_handler = 'Angles'

    @pytest.mark.parametrize("valence_handler", ["Bonds", "Angles", "ProperTorsions"])
    def test_reassign_valence(sage, valence_handler):
        """Sanity check that re-assigning valence parameters actually changes energies."""
        topology = MoleculeWithConformer.from_smiles(
            "c1c(cccc1)C(c2ccccc2)c3ccccc3",
            allow_undefined_stereo=True,
        ).to_topology()
        original_interchange = sage.create_interchange(topology)

        original_energy = get_openmm_energies(original_interchange)

        for potential in original_interchange[valence_handler].potentials.values():
            potential.parameters["k"] = potential.parameters["k"] * 100

        new_energy = get_openmm_energies(original_interchange)

>       assert original_energy != new_energy, "Energies should be different after modifying valence parameters"
E       AssertionError: Energies should be different after modifying valence parameters
E       assert EnergyReport(energies={'Bond': <Quantity(5.35485713, 'kilojoule / mole')>, 'Angle': <Quantity(29.1102705, 'kilojoule / mole')>, 'Torsion': <Quantity(20.1328021, 'kilojoule / mole')>, 'Nonbonded': <Quantity(114.818299, 'kilojoule / mole')>}) != EnergyReport(energies={'Bond': <Quantity(5.35485713, 'kilojoule / mole')>, 'Angle': <Quantity(29.1102705, 'kilojoule / mole')>, 'Torsion': <Quantity(20.1328021, 'kilojoule / mole')>, 'Nonbonded': <Quantity(114.818299, 'kilojoule / mole')>})

/Users/mattthompson/software/openff-interchange/openff/interchange/_tests/test_issues.py:295: AssertionError
______________________________________________________________ test_reassign_valence[Bonds] ______________________________________________________________

sage = <openff.toolkit.typing.engines.smirnoff.forcefield.ForceField object at 0x17a8c33b0>, valence_handler = 'Bonds'

    @pytest.mark.parametrize("valence_handler", ["Bonds", "Angles", "ProperTorsions"])
    def test_reassign_valence(sage, valence_handler):
        """Sanity check that re-assigning valence parameters actually changes energies."""
        topology = MoleculeWithConformer.from_smiles(
            "c1c(cccc1)C(c2ccccc2)c3ccccc3",
            allow_undefined_stereo=True,
        ).to_topology()
        original_interchange = sage.create_interchange(topology)

        original_energy = get_openmm_energies(original_interchange)

        for potential in original_interchange[valence_handler].potentials.values():
            potential.parameters["k"] = potential.parameters["k"] * 100

        new_energy = get_openmm_energies(original_interchange)

>       assert original_energy != new_energy, "Energies should be different after modifying valence parameters"
E       AssertionError: Energies should be different after modifying valence parameters
E       assert EnergyReport(energies={'Bond': <Quantity(5.35485713, 'kilojoule / mole')>, 'Angle': <Quantity(29.1102705, 'kilojoule / mole')>, 'Torsion': <Quantity(20.1328021, 'kilojoule / mole')>, 'Nonbonded': <Quantity(114.818299, 'kilojoule / mole')>}) != EnergyReport(energies={'Bond': <Quantity(5.35485713, 'kilojoule / mole')>, 'Angle': <Quantity(29.1102705, 'kilojoule / mole')>, 'Torsion': <Quantity(20.1328021, 'kilojoule / mole')>, 'Nonbonded': <Quantity(114.818299, 'kilojoule / mole')>})

/Users/mattthompson/software/openff-interchange/openff/interchange/_tests/test_issues.py:295: AssertionError
_________________________________________________________ test_reassign_valence[ProperTorsions] __________________________________________________________

sage = <openff.toolkit.typing.engines.smirnoff.forcefield.ForceField object at 0x17a93f860>, valence_handler = 'ProperTorsions'

    @pytest.mark.parametrize("valence_handler", ["Bonds", "Angles", "ProperTorsions"])
    def test_reassign_valence(sage, valence_handler):
        """Sanity check that re-assigning valence parameters actually changes energies."""
        topology = MoleculeWithConformer.from_smiles(
            "c1c(cccc1)C(c2ccccc2)c3ccccc3",
            allow_undefined_stereo=True,
        ).to_topology()
        original_interchange = sage.create_interchange(topology)

        original_energy = get_openmm_energies(original_interchange)

        for potential in original_interchange[valence_handler].potentials.values():
            potential.parameters["k"] = potential.parameters["k"] * 100

        new_energy = get_openmm_energies(original_interchange)

>       assert original_energy != new_energy, "Energies should be different after modifying valence parameters"
E       AssertionError: Energies should be different after modifying valence parameters
E       assert EnergyReport(energies={'Bond': <Quantity(5.35485713, 'kilojoule / mole')>, 'Angle': <Quantity(29.1102705, 'kilojoule / mole')>, 'Torsion': <Quantity(20.1328021, 'kilojoule / mole')>, 'Nonbonded': <Quantity(114.818299, 'kilojoule / mole')>}) != EnergyReport(energies={'Bond': <Quantity(5.35485713, 'kilojoule / mole')>, 'Angle': <Quantity(29.1102705, 'kilojoule / mole')>, 'Torsion': <Quantity(20.1328021, 'kilojoule / mole')>, 'Nonbonded': <Quantity(114.818299, 'kilojoule / mole')>})

/Users/mattthompson/software/openff-interchange/openff/interchange/_tests/test_issues.py:295: AssertionError
==================================================================== warnings summary ====================================================================
<frozen importlib._bootstrap>:488
<frozen importlib._bootstrap>:488
<frozen importlib._bootstrap>:488
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

<frozen importlib._bootstrap>:488
<frozen importlib._bootstrap>:488
<frozen importlib._bootstrap>:488
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
<frozen importlib._bootstrap>:488
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type swigvarlink has no __module__ attribute

openff/interchange/_tests/test_issues.py::test_reassign_valence[Angles]
openff/interchange/_tests/test_issues.py::test_reassign_vdw
openff/interchange/_tests/test_issues.py::test_reassign_valence[Bonds]
openff/interchange/_tests/test_issues.py::test_reassign_valence[ProperTorsions]
  /Users/mattthompson/software/openff-interchange/openff/interchange/components/interchange.py:1118: FutureWarning: `torch.distributed.reduce_op` is deprecated, please use `torch.distributed.ReduceOp` instead
    if isinstance(obj, functools._lru_cache_wrapper) and obj.__module__.startswith("openff.interchange"):

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
===================================================================== tests coverage =====================================================================
___________________________________________________ coverage: platform darwin, python 3.12.12-final-0 ____________________________________________________

Coverage XML written to file coverage.xml
================================================================ short test summary info =================================================================
FAILED openff/interchange/_tests/test_issues.py::test_reassign_valence[Angles] - AssertionError: Energies should be different after modifying valence parameters
FAILED openff/interchange/_tests/test_issues.py::test_reassign_valence[Bonds] - AssertionError: Energies should be different after modifying valence parameters
FAILED openff/interchange/_tests/test_issues.py::test_reassign_valence[ProperTorsions] - AssertionError: Energies should be different after modifying valence parameters
3 failed, 1 passed, 15 warnings in 5.78s
```
The output is too verbose despite `-q` - basically reiterating that the energies do **not** differ when they would be anticipated to be non-equal.

Note that the vdW parameter updates _do_ cause energies to differ. This is probably because https://github.com/openforcefield/openff-interchange/pull/1122 only caches lookups for valence handler.

**Software versions**

Reproduced using Interchange 0.5.1 - this is affected from 0.4.1 onward (https://github.com/openforcefield/openff-interchange/pull/1122)

--- a/docs/releasehistory.md
+++ b/docs/releasehistory.md
@@ -11,6 +11,10 @@ Dates are given in YYYY-MM-DD format.
 
 Please note that all releases prior to a version 1.0.0 are considered pre-releases and many API changes will come before a stable release.
 
+## Current development
+
+* #1452 Better manage some caches by invalidate OpenMM export-specific caches at export time
+
 ## 0.5.2 - 2025-03-02
 
 ### Bug fixes

--- a/openff/interchange/_tests/test_issues.py
+++ b/openff/interchange/_tests/test_issues.py
@@ -266,7 +266,7 @@ def test_issue_1433_openmm(sage, valence_handler):
 def test_issue_1450_reassign_vdw(sage, water_dimer, engine):
     """Sanity check that re-assigning vdW parameters actually changes energies."""
 
-    if not has_executable("sander"):
+    if engine == "amber" and not has_executable("sander"):
         pytest.skip(reason="sander not installed")
 
     pytest.importorskip("openmm")
@@ -299,6 +299,9 @@ def test_issue_1450_reassign_vdw(sage, water_dimer, engine):
 @pytest.mark.parametrize("valence_handler", ["Bonds", "Angles", "ProperTorsions"])
 def test_issue_1450_reassign_valence(sage, valence_handler, engine):
     """Sanity check that re-assigning valence parameters actually changes energies."""
+
+    if engine == "amber" and not has_executable("sander"):
+        pytest.skip(reason="sander not installed")
 
     topology = MoleculeWithConformer.from_smiles(
         "CC(=O)NC",

--- a/openff/interchange/_tests/test_issues.py
+++ b/openff/interchange/_tests/test_issues.py
@@ -262,6 +262,39 @@ def test_issue_1433_openmm(sage, valence_handler):
     assert abs(original_energy - new_energy) > 0.1
 
 
+def test_isssue_1450_reassign_vdw(sage, water_dimer):
+    """Sanity check that re-assigning vdW parameters actually changes energies."""
+    original_interchange = sage.create_interchange(water_dimer)
+
+    original_energy = get_openmm_energies(original_interchange).energies["Nonbonded"]
+
+    for potential in original_interchange["vdW"].potentials.values():
+        potential.parameters["epsilon"] = potential.parameters["epsilon"] * 100
+
+    new_energy = get_openmm_energies(original_interchange).energies["Nonbonded"]
+
+    assert original_energy != new_energy, "Energies should be different after modifying vdW parameters"
+
+
+@pytest.mark.parametrize("valence_handler", ["Bonds", "Angles", "ProperTorsions"])
+def test_issue_1450_reassign_valence(sage, valence_handler):
+    """Sanity check that re-assigning valence parameters actually changes energies."""
+    topology = MoleculeWithConformer.from_smiles(
+        "c1c(cccc1)C(c2ccccc2)c3ccccc3",
+        allow_undefined_stereo=True,
+    ).to_topology()
+    original_interchange = sage.create_interchange(topology)
+
+    original_energy = get_openmm_energies(original_interchange)
+
+    for potential in original_interchange[valence_handler].potentials.values():
+        potential.parameters["k"] = potential.parameters["k"] * 100
+
+    new_energy = get_openmm_energies(original_interchange)
+
+    assert original_energy != new_energy, "Energies should be different after modifying valence parameters"
+
+
 @pytest.mark.parametrize("valence_handler", ["Bonds", "Angles", "ProperTorsions", "ImproperTorsions"])
 def test_issue_1234_openmm(sage, valence_handler):
     """Test that modifications to a `ForceField` object are reflected in re-creating new `Interchange`s."""

--- a/openff/interchange/_tests/test_issues.py
+++ b/openff/interchange/_tests/test_issues.py
@@ -10,7 +10,7 @@ from openff.utilities import get_data_file_path, has_executable, skip_if_missing
 from openff.interchange import Interchange
 from openff.interchange._tests import MoleculeWithConformer, needs_gmx, shuffle_topology
 from openff.interchange.components._packmol import UNIT_CUBE, pack_box, solvate_topology
-from openff.interchange.drivers import get_amber_energies, get_openmm_energies
+from openff.interchange.drivers import get_amber_energies, get_gromacs_energies, get_openmm_energies
 from openff.interchange.exceptions import NonperiodicNoCutoffNotSupportedError
 
 
@@ -262,39 +262,84 @@ def test_issue_1433_openmm(sage, valence_handler):
     assert abs(original_energy - new_energy) > 0.1
 
 
-def test_isssue_1450_reassign_vdw(sage, water_dimer):
+@pytest.mark.parametrize("engine", ["openmm", "amber", "gromacs"])
+def test_issue_1450_reassign_vdw(sage, water_dimer, engine):
     """Sanity check that re-assigning vdW parameters actually changes energies."""
+
+    if not has_executable("sander"):
+        pytest.skip(reason="sander not installed")
+
+    pytest.importorskip("openmm")
+
     original_interchange = sage.create_interchange(water_dimer)
 
-    original_energy = get_openmm_energies(original_interchange).energies["Nonbonded"]
+    match engine:
+        case "openmm":
+            original_energy = get_openmm_energies(original_interchange, combine_nonbonded_forces=False).energies["vdW"]
+        case "amber":
+            original_energy = get_amber_energies(original_interchange).energies["vdW"]
+        case "gromacs":
+            original_energy = get_gromacs_energies(original_interchange).energies["vdW"]
 
     for potential in original_interchange["vdW"].potentials.values():
         potential.parameters["epsilon"] = potential.parameters["epsilon"] * 100
 
-    new_energy = get_openmm_energies(original_interchange).energies["Nonbonded"]
+    match engine:
+        case "openmm":
+            new_energy = get_openmm_energies(original_interchange, combine_nonbonded_forces=False).energies["vdW"]
+        case "amber":
+            new_energy = get_amber_energies(original_interchange).energies["vdW"]
+        case "gromacs":
+            new_energy = get_gromacs_energies(original_interchange).energies["vdW"]
 
     assert original_energy != new_energy, "Energies should be different after modifying vdW parameters"
 
 
+@pytest.mark.parametrize("engine", ["openmm", "amber", "gromacs"])
 @pytest.mark.parametrize("valence_handler", ["Bonds", "Angles", "ProperTorsions"])
-def test_issue_1450_reassign_valence(sage, valence_handler):
+def test_issue_1450_reassign_valence(sage, valence_handler, engine):
     """Sanity check that re-assigning valence parameters actually changes energies."""
+
     topology = MoleculeWithConformer.from_smiles(
-        "c1c(cccc1)C(c2ccccc2)c3ccccc3",
+        "CC(=O)NC",
         allow_undefined_stereo=True,
     ).to_topology()
+
+    topology.box_vectors = Quantity(numpy.eye(3) * 10.0, "nanometer")
+
     original_interchange = sage.create_interchange(topology)
 
-    original_energy = get_openmm_energies(original_interchange)
+    match valence_handler:
+        case "Bonds":
+            shorthand = "Bond"
+        case "Angles":
+            shorthand = "Angle"
+        case "ProperTorsions":
+            shorthand = "Torsion"
+
+    match engine:
+        case "openmm":
+            original_energy = get_openmm_energies(original_interchange).energies[shorthand]
+        case "amber":
+            original_energy = get_amber_energies(original_interchange).energies[shorthand]
+        case "gromacs":
+            original_energy = get_gromacs_energies(original_interchange).energies[shorthand]
 
     for potential in original_interchange[valence_handler].potentials.values():
         potential.parameters["k"] = potential.parameters["k"] * 100
 
-    new_energy = get_openmm_energies(original_interchange)
+    match engine:
+        case "openmm":
+            new_energy = get_openmm_energies(original_interchange).energies[shorthand]
+        case "amber":
+            new_energy = get_amber_energies(original_interchange).energies[shorthand]
+        case "gromacs":
+            new_energy = get_gromacs_energies(original_interchange).energies[shorthand]
 
     assert original_energy != new_energy, "Energies should be different after modifying valence parameters"
 
 
+@skip_if_missing("openmm")
 @pytest.mark.parametrize("valence_handler", ["Bonds", "Angles", "ProperTorsions", "ImproperTorsions"])
 def test_issue_1234_openmm(sage, valence_handler):
     """Test that modifications to a `ForceField` object are reflected in re-creating new `Interchange`s."""

--- a/openff/interchange/_tests/test_issues.py
+++ b/openff/interchange/_tests/test_issues.py
@@ -300,6 +300,8 @@ def test_issue_1450_reassign_vdw(sage, water_dimer, engine):
 def test_issue_1450_reassign_valence(sage, valence_handler, engine):
     """Sanity check that re-assigning valence parameters actually changes energies."""
 
+    pytest.importorskip("openmm")
+
     if engine == "amber" and not has_executable("sander"):
         pytest.skip(reason="sander not installed")
 
@@ -330,6 +332,9 @@ def test_issue_1450_reassign_valence(sage, valence_handler, engine):
 
     for potential in original_interchange[valence_handler].potentials.values():
         potential.parameters["k"] = potential.parameters["k"] * 100
+
+        if "phase" in potential.parameters:
+            potential.parameters["phase"] = potential.parameters["phase"] + Quantity(20, "degree")
 
     match engine:
         case "openmm":

--- a/openff/interchange/interop/openmm/_valence.py
+++ b/openff/interchange/interop/openmm/_valence.py
@@ -77,8 +77,6 @@ def _process_bond_forces(
     """
     Process the Bonds section of an Interchange object.
     """
-    _lookup_bond_parameters_from_potential_key.cache_clear()
-
     try:
         bond_handler = interchange.collections["Bonds"]
     except KeyError:
@@ -140,8 +138,6 @@ def _process_angle_forces(
     """
     Process the Angles section of an Interchange object.
     """
-    _lookup_angle_parameters_from_potential_key.cache_clear()
-
     try:
         angle_handler = interchange.collections["Angles"]
     except KeyError:
@@ -237,8 +233,6 @@ def _process_proper_torsion_forces(interchange, openmm_sys, particle_map):
     """
     Process the Propers section of an Interchange object.
     """
-    _lookup_torsion_parameters_from_potential_key.cache_clear()
-
     torsion_force = openmm.PeriodicTorsionForce()
     openmm_sys.addForce(torsion_force)
 
@@ -304,8 +298,6 @@ def _process_improper_torsion_forces(interchange, openmm_sys, particle_map):
     """
     Process the Impropers section of an Interchange object.
     """
-    _lookup_torsion_parameters_from_potential_key.cache_clear()
-
     if "ImproperTorsions" not in interchange.collections.keys():
         return
 

--- a/openff/interchange/interop/openmm/_valence.py
+++ b/openff/interchange/interop/openmm/_valence.py
@@ -77,6 +77,8 @@ def _process_bond_forces(
     """
     Process the Bonds section of an Interchange object.
     """
+    _lookup_bond_parameters_from_potential_key.cache_clear()
+
     try:
         bond_handler = interchange.collections["Bonds"]
     except KeyError:
@@ -138,6 +140,8 @@ def _process_angle_forces(
     """
     Process the Angles section of an Interchange object.
     """
+    _lookup_angle_parameters_from_potential_key.cache_clear()
+
     try:
         angle_handler = interchange.collections["Angles"]
     except KeyError:
@@ -233,6 +237,8 @@ def _process_proper_torsion_forces(interchange, openmm_sys, particle_map):
     """
     Process the Propers section of an Interchange object.
     """
+    _lookup_torsion_parameters_from_potential_key.cache_clear()
+
     torsion_force = openmm.PeriodicTorsionForce()
     openmm_sys.addForce(torsion_force)
 
@@ -298,6 +304,8 @@ def _process_improper_torsion_forces(interchange, openmm_sys, particle_map):
     """
     Process the Impropers section of an Interchange object.
     """
+    _lookup_torsion_parameters_from_potential_key.cache_clear()
+
     if "ImproperTorsions" not in interchange.collections.keys():
         return
 

--- a/openff/interchange/interop/openmm/_valence.py
+++ b/openff/interchange/interop/openmm/_valence.py
@@ -304,6 +304,8 @@ def _process_improper_torsion_forces(interchange, openmm_sys, particle_map):
     """
     Process the Impropers section of an Interchange object.
     """
+    # Possible performance improvement to be had by more carefully managing this cache, see:
+    # https://github.com/openforcefield/openff-interchange/pull/1452#discussion_r2891781038
     _lookup_torsion_parameters_from_potential_key.cache_clear()
 
     if "ImproperTorsions" not in interchange.collections.keys():

